### Homework 7

##### Exercise 3
##### a) Explain how k-fold cross-validation is implemented.

The k-fold cross validation approach divides the original data set into k groups about the same size. One group is treated as a test data set and the other k-1 groups are used as a training data set. The model will be fit on the training set then computed on the test set, so one can estimate how well the model fits on new data.

##### b) What are the advantages and disadvantages of k-fold crossvalidation relative to:
##### i. The validation set approach?
##### ii. LOOCV?

b) i. Test error rates of a k-fold CV are not highly variable as with the validation set approach, since the validation set approach depends on which observation group is used as the training and test data set. Also it tends to overestimate the test error rate of the model fit on the whole data set. 

ii. The k-fold CV has computational advantages to LOOCV, since the latter can be time consuming depending on the sample size. Another advantage lies in bias-variance trade-off:k-fold CV using k=5 or k=10 don't suffer from high bias or high variance.

##### Exercise 7

##### (a) Fit a logistic regression model that predicts Direction using Lag1 and Lag2.

In [1]:
library(ISLR)

In [2]:
head(Weekly)
attach(Weekly)
Weekly

Year Lag1   Lag2   Lag3   Lag4   Lag5   Volume    Today  Direction
1 1990  0.816  1.572 -3.936 -0.229 -3.484 0.1549760 -0.270 Down     
2 1990 -0.270  0.816  1.572 -3.936 -0.229 0.1485740 -2.576 Down     
3 1990 -2.576 -0.270  0.816  1.572 -3.936 0.1598375  3.514 Up       
4 1990  3.514 -2.576 -0.270  0.816  1.572 0.1616300  0.712 Up       
5 1990  0.712  3.514 -2.576 -0.270  0.816 0.1537280  1.178 Up       
6 1990  1.178  0.712  3.514 -2.576 -0.270 0.1544440 -1.372 Down

Year Lag1   Lag2   Lag3   Lag4   Lag5   Volume    Today  Direction
1    1990  0.816  1.572 -3.936 -0.229 -3.484 0.1549760 -0.270 Down     
2    1990 -0.270  0.816  1.572 -3.936 -0.229 0.1485740 -2.576 Down     
3    1990 -2.576 -0.270  0.816  1.572 -3.936 0.1598375  3.514 Up       
4    1990  3.514 -2.576 -0.270  0.816  1.572 0.1616300  0.712 Up       
5    1990  0.712  3.514 -2.576 -0.270  0.816 0.1537280  1.178 Up       
6    1990  1.178  0.712  3.514 -2.576 -0.270 0.1544440 -1.372 Down     
7    1990 -1.372  1.178  0.712  3.514 -2.576 0.1517220  0.807 Up       
8    1990  0.807 -1.372  1.178  0.712  3.514 0.1323100  0.041 Up       
9    1990  0.041  0.807 -1.372  1.178  0.712 0.1439720  1.253 Up       
10   1990  1.253  0.041  0.807 -1.372  1.178 0.1336350 -2.678 Down     
11   1990 -2.678  1.253  0.041  0.807 -1.372 0.1490240 -1.793 Down     
12   1990 -1.793 -2.678  1.253  0.041  0.807 0.1357900  2.820 Up       
13   1990  2.820 -1.793 -2.678  1.253  0.041 0.1398980  4.022 Up       
14   1990  4.022  2.820 -1.793 -2.678  1.253 0.1643420  0.750 Up       
15   1990  0.750  4.022  2.820 -1.793 -2.678 0.1756480 -0.017 Down     
16   1990 -0.017  0.750  4.022  2.820 -1.793 0.1634700  2.420 Up       
17   1990  2.420 -0.017  0.750  4.022  2.820 0.1726250 -1.225 Down     
18   1990 -1.225  2.420 -0.017  0.750  4.022 0.1684460  1.171 Up       
19   1990  1.171 -1.225  2.420 -0.017  0.750 0.1552920 -2.061 Down     
20   1990 -2.061  1.171 -1.225  2.420 -0.017 0.1433920  0.729 Up       
21   1990  0.729 -2.061  1.171 -1.225  2.420 0.1405540  0.112 Up       
22   1990  0.112  0.729 -2.061  1.171 -1.225 0.1250750  2.480 Up       
23   1990  2.480  0.112  0.729 -2.061  1.171 0.1716040 -1.552 Down     
24   1990 -1.552  2.480  0.112  0.729 -2.061 0.1669560 -2.259 Down     
25   1990 -2.259 -1.552  2.480  0.112  0.729 0.1717180 -2.428 Down     
26   1990 -2.428 -2.259 -1.552  2.480  0.112 0.2098160 -2.708 Down     
27   1990 -2.708 -2.428 -2.259 -1.552  2.480 0.1927060 -2.292 Down     
28   1990 -2.292 -2.708 -2.428 -2.259 -1.552 0.1482520 -4.978 Down     
29   1990 -4.978 -2.292 -2.708 -2.428 -2.259 0.1898580  3.547 Up       
30   1990  3.547 -4.978 -2.292 -2.708 -2.428 0.1278840  0.260 Up       
...  ...  ...    ...    ...    ...    ...    ...       ...    ...      
1060 2010 -2.252  0.158 -4.226  2.232 -6.388 5.368597   2.509 Up       
1061 2010  2.509 -2.252  0.158 -4.226  2.232 5.369514   2.374 Up       
1062 2010  2.374  2.509 -2.252  0.158 -4.226 4.637208  -3.646 Down     
1063 2010 -3.646  2.374  2.509 -2.252  0.158 4.699712  -5.032 Down     
1064 2010 -5.032 -3.646  2.374  2.509 -2.252 5.100892   5.416 Up       
1065 2010  5.416 -5.032 -3.646  2.374  2.509 4.419372  -1.213 Down     
1066 2010 -1.213  5.416 -5.032 -3.646  2.374 4.487664   3.548 Up       
1067 2010  3.548 -1.213  5.416 -5.032 -3.646 4.580286  -0.096 Down     
1068 2010 -0.096  3.548 -1.213  5.416 -5.032 4.271320   1.819 Up       
1069 2010  1.819 -0.096  3.548 -1.213  5.416 3.963460  -3.779 Down     
1070 2010 -3.779  1.819 -0.096  3.548 -1.213 3.906558  -0.700 Down     
1071 2010 -0.700 -3.779  1.819 -0.096  3.548 3.777406  -0.663 Down     
1072 2010 -0.663 -0.700 -3.779  1.819 -0.096 3.951328   3.750 Up       
1073 2010  3.750 -0.663 -0.700 -3.779  1.819 3.718470   0.456 Up       
1074 2010  0.456  3.750 -0.663 -0.700 -3.779 3.195238   1.446 Up       
1075 2010  1.446  0.456  3.750 -0.663 -0.700 3.972432   2.050 Up       
1076 2010  2.050  1.446  0.456  3.750 -0.663 3.884522  -0.212 Down     
1077 2010 -0.212  2.050  1.446  0.456  3.750 4.037410   1.650 Up       
1078 2010  1.650 -0.212  2.050  1.446  0.456 3.905616   0.948 Up       
1079 2010  0.948  1.650 -0.212  2.050  1.446 4.449160   0.586 Up       
1080 2010  0.586  0.948  1.650 -0.212  2.050 4.576282   0.015 Up       
1081 2010  0.015  0.586  0.948  1.650 -0.212 4.116414   3.599 Up       
1082 2010  3.599  0.015  0.586  0.948  1.650 4.798758  -2.173 Down     
1083 2010 -2.173  3.599  0.015  0.586  0.948 

In [3]:
glm1 <-glm(Direction ~ Lag1 + Lag2, data = Weekly, family = binomial)
summary(glm1)


Call:
glm(formula = Direction ~ Lag1 + Lag2, family = binomial, data = Weekly)

Deviance Residuals: 
   Min      1Q  Median      3Q     Max  
-1.623  -1.261   1.001   1.083   1.506  

Coefficients:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept)  0.22122    0.06147   3.599 0.000319 ***
Lag1        -0.03872    0.02622  -1.477 0.139672    
Lag2         0.06025    0.02655   2.270 0.023232 *  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1496.2  on 1088  degrees of freedom
Residual deviance: 1488.2  on 1086  degrees of freedom
AIC: 1494.2

Number of Fisher Scoring iterations: 4


##### (b) Fit a logistic regression model that predicts Direction using Lag1 and Lag2 using all but the first observation.

In [4]:
glm2 <-glm(Direction ~ Lag1 + Lag2, data = Weekly[-1, ], family = binomial)
summary(glm2)


Call:
glm(formula = Direction ~ Lag1 + Lag2, family = binomial, data = Weekly[-1, 
    ])

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-1.6258  -1.2617   0.9999   1.0819   1.5071  

Coefficients:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept)  0.22324    0.06150   3.630 0.000283 ***
Lag1        -0.03843    0.02622  -1.466 0.142683    
Lag2         0.06085    0.02656   2.291 0.021971 *  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1494.6  on 1087  degrees of freedom
Residual deviance: 1486.5  on 1085  degrees of freedom
AIC: 1492.5

Number of Fisher Scoring iterations: 4


##### (c) Use the model from (b) to predict the direction of the first observation. You can do this by predicting that the first observation will go up if P(Direction="Up"|Lag1, Lag2) > 0.5. Was this observation correctly classified?

In [5]:
glm.pred <-predict(glm2, Weekly[1, ], type = 'response')
ifelse(glm.pred > 0.5, 'Up')
glm.pred

1 
"Up"

1 
0.5713923

The value 0.5713 is above 0.5 and is correctly classified as 'Up'.

##### (d) Write a for loop from i = 1 to i = n, where n is the number of observations in the data set, that performs each of the following steps:
##### i. Fit a logistic regression model using all but the ith observation to predict Direction using Lag1 and Lag2.
##### ii. Compute the posterior probability of the market moving up for the ith observation.
##### iii. Use the posterior probability for the ith observation in order to predict whether or not the market moves up.
##### iv. Determine whether or not an error was made in predicting the direction for the ith observation. If an error was   made, then indicate this as a 1, and otherwise indicate it as a 0.

In [6]:
dim(Weekly)

[1] 1089    9

In [7]:
error <- c(1, 1089)
for(i in 1:1089) {
    glm.fit <-glm(Direction~Lag1 +Lag2, data=Weekly[-i, ], family = binomial)
    glm.pred <-ifelse(predict(glm.fit, Weekly[i, ], type = 'response') > 0.5, 'Up', 'Down')
    error[i] <-if (glm.pred != Weekly[i,'Direction']) 1 else 0
                 }
error

[1] 1 1 0 1 0 1 0 0 0 1 1 0 1 0 1 0 1 0 1 0 0 0 1 1 1 1 1 1 0 1 1 1 1 0 1 0 0
  [38] 0 1 0 1 0 0 1 0 1 1 1 0 1 0 0 0 1 0 0 1 1 0 0 0 0 1 0 1 1 0 0 1 0 1 1 0 0
  [75] 0 1 0 1 1 0 0 1 1 0 1 1 0 0 1 0 0 1 1 1 0 0 0 0 0 1 0 1 1 0 0 1 0 1 0 0 1
 [112] 1 0 0 1 0 0 1 0 0 1 1 1 1 0 0 0 1 0 1 0 1 1 0 0 0 1 1 1 0 0 0 1 0 0 0 0 0
 [149] 0 1 1 1 0 1 0 0 1 1 0 1 0 0 1 1 0 0 1 0 0 1 0 0 1 1 1 0 1 0 1 0 0 0 0 0 0
 [186] 0 0 1 1 0 1 0 1 0 1 0 1 0 0 1 0 0 1 0 0 1 0 1 0 1 1 1 0 0 1 1 0 1 0 0 1 1
 [223] 0 0 0 1 1 1 0 1 0 1 0 1 0 0 0 1 1 0 1 0 1 0 1 0 1 0 1 1 0 1 0 0 1 0 0 1 0
 [260] 0 0 0 0 1 0 0 0 1 0 0 1 0 0 0 1 0 0 1 0 0 1 0 0 1 0 1 1 0 0 0 0 0 1 0 1 0
 [297] 0 1 0 0 0 1 0 0 1 1 0 0 1 0 0 0 0 1 0 1 1 0 0 1 0 1 0 1 1 0 0 0 1 0 1 0 0
 [334] 1 1 1 1 0 1 0 0 1 0 0 0 1 0 1 0 1 0 0 0 0 0 1 1 0 0 1 0 0 1 0 0 0 1 1 0 1
 [371] 1 1 1 1 0 0 0 1 0 0 0 0 0 0 1 0 1 1 0 0 1 1 0 0 0 0 0 1 0 0 1 1 1 0 1 0 1
 [408] 0 1 1 1 0 1 0 0 0 0 0 0 0 0 0 0 1 0 1 0 1 0 1 0 0 1 0 1 0 0 0 0 0 1 1 1 1
 [445] 0 1 1 0 1 0 1 1 0 1 0 0 1 0 0 1 1 0 0 0 0 1 1 0 0 1 0 1 0 0 0 1 0 0 1 0 0
 [482] 0 1 1 1 0 1 0 0 0 1 0 1 1 1 0 0 0 0 1 1 1 0 1 1 0 1 0 0 0 1 0 1 0 0 0 1 0
 [519] 1 1 0 0 1 1 0 0 0 1 1 0 1 0 1 1 1 1 1 0 0 0 1 0 0 0 1 1 0 1 0 0 0 1 1 1 1
 [556] 1 1 0 1 0 1 0 0 1 0 0 1 1 1 0 0 0 1 1 1 1 1 1 1 1 1 1 0 1 0 0 1 0 0 1 0 1
 [593] 0 0 1 0 0 1 0 1 1 0 1 1 1 0 1 0 1 0 1 0 0 0 1 0 1 0 1 0 1 1 0 1 1 0 1 0 1
 [630] 0 1 1 1 1 0 1 1 0 0 0 1 1 1 1 0 1 1 1 0 1 0 0 0 1 1 1 1 1 1 0 1 0 0 1 0 0
 [667] 0 1 1 0 1 0 1 1 1 1 0 0 0 1 1 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 0 0 1 0 1 1
 [704] 0 0 0 0 1 0 1 0 1 0 1 0 0 1 1 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 1 1 1 0 1 1 0
 [741] 1 1 1 1 0 0 0 1 1 1 1 1 1 0 1 0 0 0 0 0 0 1 0 1 1 1 0 0 0 1 0 0 1 0 0 0 1
 [778] 1 1 0 0 0 1 0 0 1 1 1 0 0 1 0 1 0 1 0 0 1 0 0 1 0 0 0 0 0 1 0 1 1 0 0 1 1
 [815] 0 1 1 1 0 0 0 0 0 1 1 0 0 1 0 0 1 0 1 0 0 0 1 1 0 1 1 0 1 0 1 0 1 1 0 0 1
 [852] 1 1 0 1 1 0 0 0 1 0 1 0 1 0 1 0 0 0 0 0 1 0 0 1 1 0 0 1 0 1 0 1 1 0 1 0 1
 [889] 1 0 0 0 1 0 0 0 0 0 0 0 1 0 1 0 1 0 0 0 1 1 1 1 1 0 1 1 0 0 0 0 0 1 0 0 1
 [926] 0 0 0 0 1 0 1 1 1 0 0 1 1 0 1 1 1 1 0 1 0 1 0 1 0 1 0 1 0 0 1 1 1 1 1 0 1
 [963] 0 0 0 1 1 1 0 1 1 1 1 0 0 0 0 1 1 0 0 0 0 1 0 0 1 1 1 0 0 1 1 1 0 1 0 0 0
[1000] 0 1 0 0 1 0 1 0 0 1 1 1 1 0 1 0 0 1 0 0 1 0 0 1 1 0 1 1 1 0 1 1 0 0 0 1 0
[1037] 1 0 1 1 1 1 0 0 1 0 0 0 0 0 0 1 0 1 1 0 0 0 0 0 1 1 1 0 0 0 1 0 1 1 1 0 0
[1074] 0 0 1 0 0 0 0 0 1 0 1 0 0 0 0 0

##### (e) Take the average of the n numbers obtained in (d)iv in order to obtain the LOOCV estimate for the test error. Comment on the results.

In [8]:
mean(error)

[1] 0.4499541

The test error rate is 44.99%.

##### Exercise 9

##### (a) Based on this data set, provide an estimate for the population mean of medv. Call this estimate ˆμ.

In [9]:
library(MASS)

In [10]:
head(Boston)
attach(Boston)

crim    zn indus chas nox   rm    age  dis    rad tax ptratio black  lstat
1 0.00632 18 2.31  0    0.538 6.575 65.2 4.0900 1   296 15.3    396.90 4.98 
2 0.02731  0 7.07  0    0.469 6.421 78.9 4.9671 2   242 17.8    396.90 9.14 
3 0.02729  0 7.07  0    0.469 7.185 61.1 4.9671 2   242 17.8    392.83 4.03 
4 0.03237  0 2.18  0    0.458 6.998 45.8 6.0622 3   222 18.7    394.63 2.94 
5 0.06905  0 2.18  0    0.458 7.147 54.2 6.0622 3   222 18.7    396.90 5.33 
6 0.02985  0 2.18  0    0.458 6.430 58.7 6.0622 3   222 18.7    394.12 5.21 
  medv
1 24.0
2 21.6
3 34.7
4 33.4
5 36.2
6 28.7

In [11]:
dim(Boston)

[1] 506  14

In [12]:
û <-mean(medv)
û

[1] 22.53281

##### (b) Provide an estimate of the standard error of ˆ μ. Interpret this result.
Hint: We can compute the standard error of the sample mean by dividing the sample standard deviation by the square root of the number of observations.

In [13]:
se <-sd(medv)/sqrt(length(medv))
se

[1] 0.4088611

The standard error of 0.40886 tells us, how much the mean of the sample can vary compared to the mean of the true population.

##### (c) Now estimate the standard error of ˆ μ using the bootstrap. How does this compare to your answer from (b)?

In [14]:
library(boot)

In [15]:
set.seed(1)
boot1 <- function(data,index) {
    x <- mean(data[index])
    return(x)
}
boot(medv, boot1, R=1000 )


ORDINARY NONPARAMETRIC BOOTSTRAP


Call:
boot(data = medv, statistic = boot1, R = 1000)


Bootstrap Statistics :
    original      bias    std. error
t1* 22.53281 0.007650791   0.4106622

The standard error estimate from the bootstrap is similar to the estimate from (b): 0.410 vs. 0.408.

##### (d) Based on your bootstrap estimate from (c), provide a 95 % confidence interval for the mean of medv. Compare it to the results obtained using t.test(Boston$medv).
Hint: You can approximate a 95 % confidence interval using the
formula [ˆ μ − 2SE(ˆ μ), μ ˆ + 2SE(ˆ μ)].

In [16]:
CIplus <- c(û + 2 * se)
CIplus
CIminus <-c(û - 2 * se)
CIminus

[1] 23.35053

[1] 21.71508

In [17]:
t.test(Boston$medv)


	One Sample t-test

data:  Boston$medv
t = 55.111, df = 505, p-value < 2.2e-16
alternative hypothesis: true mean is not equal to 0
95 percent confidence interval:
 21.72953 23.33608
sample estimates:
mean of x 
 22.53281 


The CI's are similar.

##### (e) Based on this data set, provide an estimate, ˆ μmed, for the median value of medv in the population.

In [18]:
med <-median(medv)
med

[1] 21.2

##### (f) We now would like to estimate the standard error of ˆ μmed. Unfortunately, there is no simple formula for computing the standard error of the median. Instead, estimate the standard error of the median using the bootstrap. Comment on your findings.

In [19]:
set.seed(1)
boot2 <- function(data,index) {
    x <- median(data[index])
    return(x)
}
boot(medv, boot2, R=1000 )


ORDINARY NONPARAMETRIC BOOTSTRAP


Call:
boot(data = medv, statistic = boot2, R = 1000)


Bootstrap Statistics :
    original  bias    std. error
t1*     21.2 0.02295   0.3778075

The bootstrap gives the same output for median (21.2) and the standard error of the median is 0.3778.

##### (g) Based on this data set, provide an estimate for the tenth percentile of medv in Boston suburbs. Call this quantity ˆ μ0.1. (You can use the quantile() function.)

In [20]:
perc. <-quantile(medv, seq(0, 1, 0.1))
perc.
perc.tenth <-quantile(medv, (0.1))
perc.tenth

0%   10%   20%   30%   40%   50%   60%   70%   80%   90%  100% 
 5.00 12.75 15.30 18.20 19.70 21.20 22.70 24.15 28.20 34.80 50.00

10% 
12.75

##### (h) Use the bootstrap to estimate the standard error of ˆ μ0.1. Comment on your findings

In [21]:
set.seed(1)
boot3 <- function(data,index) {
    x <- quantile(data[index])
    return(x)
}
boot(medv, boot3, R=1000 )

set.seed(1)
boot2 <- function(data,index) {
    x <- quantile(data[index], (0.1))
    return(x)
}
boot(medv, boot2, R=1000 )


ORDINARY NONPARAMETRIC BOOTSTRAP


Call:
boot(data = medv, statistic = boot3, R = 1000)


Bootstrap Statistics :
    original    bias    std. error
t1*    5.000  0.106800   0.3338941
t2*   17.025 -0.059775   0.5508207
t3*   21.200  0.022950   0.3778075
t4*   25.000  0.385150   0.8758255
t5*   50.000  0.000000   0.0000000


ORDINARY NONPARAMETRIC BOOTSTRAP


Call:
boot(data = medv, statistic = boot2, R = 1000)


Bootstrap Statistics :
    original  bias    std. error
t1*    12.75  0.0339   0.4767526

Both of them give the same output of 12.75 and the SE is 0.476.